src/lib/operators/index_scan.cpp

```diff
@@ -0,0 +1,154 @@
+#include "index_scan.hpp"
+
+#include <algorithm>
+
+#include "scheduler/abstract_task.hpp"
+#include "scheduler/current_scheduler.hpp"
+#include "scheduler/job_task.hpp"
+
+#include "storage/index/base_index.hpp"
+#include "storage/reference_column.hpp"
+
+#include "utils/assert.hpp"
+
+
+namespace opossum {
+
+IndexScan::IndexScan(std::shared_ptr<AbstractOperator> in, std::vector<ChunkID> chunk_ids,
```

Comment:
I think we have not completely discussed this. In my mind, the chunk_ids were part of the `Abstract(ReadOnly)Operator` and you could either call `op->set_positive_chunk_list` or `op->set_negative_chunk_list`. I see two advantages with that:

(1) The default case, in which you scan the entire table remains unchanged
(2) It is easier to distinguish between positive and negative lists.

Reply:
The disadvantage is that it adds methods to a common interface which is only used in two sub classes and ignored in the rest. Am I wrong?

Reply:
We would also have it in the joins, where we might be able to do an indexjoin on one part and a non-index join on another.

Reply:
Then it wouldn’t be clear which input is meant. The first or the second?

Reply:
Good point. How about we move it to the end as an optional parameter? We still need to decide how to do the positive/negative thing then.

Reply:
Agreed. We could also create two constructors just as in the old days when there didn’t have `std::optional`. And I could rename the variable to `included_chunk_ids` vs. `excluded_chunk_ids`.

Reply:
How would you call it though? The semantics if it is inclusive or exclusive should not only be stored in the name of the parameter. Maybe a pair `<bool inclusive, poslist>`?

Reply:
In this case, the indexscan would be inclusive, the tablescan exclusive. I wouldn't want a parameter that has the same type but means one thing in one case and the opposite in the other.

The IndexScan should not allow you to pass an exclusive list, because you cannot be sure that all chunks that were added since you created that list have an index.

Reply:
We could still add non-virtual methods to `TableScan` and `IndexScan`. For the former: `set_excluded_chunks_ids`, for the latter: `set_included_chunk_ids`.

Reply:
Sounds good.